### PR TITLE
Add ProductVariantBulkUpdate mutation and refactor ProductVariantBulkCreate

### PR DESCRIPTION
**Pull request by SzymJ**
I want to merge this change because it adds `ProductVariantBulkUpdate` and refactors error handling and response in `ProductVariantBulkCreate` mutations related to #10798 RFC.

# Impact

* [ ] New migrations
* [x] New/Updated API fields or mutations
* [x] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/graphql/schema.graphql [hunk 1]
```diff
@@ -18482,6 +18536,13 @@ type BulkProductError {
   channels: [ID!]
 }
 
+"""An enumeration."""
+enum ErrorPolicyEnum {
+  IGNORE_FAILED
+  REJECT_EVERYTHING
+  REJECT_FAILED_ROWS
+}
```
Comment (maarcingebala) [inline saleor/graphql/schema.graphql:18544]: Can we provide descriptions for these?

File: saleor/graphql/product/bulk_mutations/product_variant_bulk_update.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+            required=False,
+            default_value=ErrorPolicyEnum.REJECT_EVERYTHING.value,
+            description="Policies of error handling.",
+        )
+
+    class Meta:
+        description = "Update a given product variants."
+        permissions = (ProductPermissions.MANAGE_PRODUCTS,)
+        error_type_class = BulkProductError
+        error_type_field = "bulk_product_errors"
```
Comment (fowczarek) [inline saleor/graphql/product/bulk_mutations/product_variant_bulk_update.py:78]: We could skip deprecated errors in new mutations. 
```suggestion
        error_type_class = BulkProductError
```

**Review by fowczarek — COMMENTED** [review]
Could you add benchmarks for new mutations?

File: saleor/graphql/product/bulk_mutations/product_variant_bulk_create.py [hunk 3]
```diff
@@ -351,57 +665,120 @@ def create_variant_channel_listings(cls, variant, cleaned_input):
...
-            product.default_variant = instances[0]
-            product.save(update_fields=["default_variant", "updated_at"])
+    def save_variants(cls, info, instances_data_with_errors_list, product):
+        for instance_data in instances_data_with_errors_list:
+            instance = instance_data["instance"]
+            if instance:
+                cleaned_input = instance_data.pop("cleaned_input")
+                cls.save(info, instance_data["instance"], cleaned_input)
+                cls.create_variant_stocks(instance, cleaned_input)
+                cls.create_variant_channel_listings(instance, cleaned_input)
```
Comment (IKarbowiak) [inline saleor/graphql/product/bulk_mutations/product_variant_bulk_create.py:675]: Couldn't we do this once for all variants?

File: saleor/graphql/product/bulk_mutations/product_variant_bulk_update.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+            required=True,
+        )
+        error_policy = ErrorPolicyEnum(
+            required=False,
+            default_value=ErrorPolicyEnum.REJECT_EVERYTHING.value,
+            description="Policies of error handling.",
+        )
+
+    class Meta:
+        description = "Update a given product variants."
```
[the remark this answers is not in the document]
Comment (IKarbowiak) [reply saleor/graphql/product/bulk_mutations/product_variant_bulk_update.py:75]: Also lack of `ADDED_IN_X` and preview feature label.

File: saleor/graphql/product/bulk_mutations/product_variant_bulk_update.py [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+from ..utils import get_used_variants_attribute_values
+from .product_variant_bulk_create import (
+    BulkAttributeValueInput,
+    ProductVariantBulkCreate,
+    ProductVariantBulkCreateInput,
+    ProductVariantBulkResult,
+)
+
+
+class ProductVariantBulkUpdateInput(ProductVariantBulkCreateInput):
```
Comment (IKarbowiak) [inline saleor/graphql/product/bulk_mutations/product_variant_bulk_update.py:30]: IMO we should add the description with the `ADDED_IN_X` label

File: saleor/graphql/product/bulk_mutations/product_variant_bulk_update.py (hunk 2 above)
Comment (IKarbowiak) [inline saleor/graphql/product/bulk_mutations/product_variant_bulk_update.py:78]: Please remove `error_type_field`. It's adding a deprecated field.

File: saleor/graphql/product/bulk_mutations/product_variant_bulk_update.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+
+            if "preorder_threshold" in channel_listing_data:
+                defaults["preorder_quantity_threshold"] = preorder_quantity_threshold
+
+            if "cost_price" in channel_listing_data:
+                defaults["cost_price_amount"] = cost_price
+
+            models.ProductVariantChannelListing.objects.update_or_create(
+                variant=variant, channel=channel, defaults=defaults
+            )
```
Comment (IKarbowiak) [inline saleor/graphql/product/bulk_mutations/product_variant_bulk_update.py:234]: I do not like this 😄 . We should take care of creation and update separately to do this in bulk

File: saleor/graphql/product/bulk_mutations/product_variant_bulk_update.py [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+    @classmethod
+    @traced_atomic_transaction()
+    def save_variants(cls, info, manager, instances_data_with_errors_list):
+        for instance_data in instances_data_with_errors_list:
+            instance = instance_data["instance"]
+            if instance:
+                cleaned_input = instance_data.pop("cleaned_input")
+                cls.save(info, instance, cleaned_input)
+                cls.update_or_create_variant_stocks(instance, cleaned_input, manager)
+                cls.update_or_create_variant_channel_listings(instance, cleaned_input)
```
Comment (IKarbowiak) [inline saleor/graphql/product/bulk_mutations/product_variant_bulk_update.py:245]: Cannot do this in bulk? We are running this as many times as many variants to update are provided. We should aim for a constant number of db queries.

File: saleor/graphql/product/tests/deprecated/test_product_variant_bulk_create.py [hunk 8]
```diff
@@ -0,0 +1,1143 @@
...
+    assert data["productVariants"][0]["name"] == attribute_value.name
+    assert product_variant_count + 1 == ProductVariant.objects.count()
+    assert attribute_value_count == size_attribute.values.count()
+    product_variant = ProductVariant.objects.get(sku=sku)
+    product.refresh_from_db()
+    assert product.default_variant == product_variant
+    assert product_variant_created_webhook_mock.call_count == data["count"]
+
+
+def test_product_variant_bulk_create_with_swatch_attribute(
```
Comment (IKarbowiak) [inline saleor/graphql/product/tests/deprecated/test_product_variant_bulk_create.py:155]: Lack of `given, when, then` parts in most of the tests. It's hard to read :(

File: saleor/graphql/product/bulk_mutations/product_variant_bulk_update.py (hunk 2 above)
Comment (fowczarek) [reply saleor/graphql/product/bulk_mutations/product_variant_bulk_update.py:78]: Same as:
https://github.com/saleor/saleor/pull/11394#discussion_r1082241993

File: saleor/graphql/product/bulk_mutations/product_variant_bulk_create.py [hunk 9]
```diff
@@ -252,83 +437,212 @@ def save(cls, info, instance, cleaned_input):
...
-        if sku in sku_list:
-            errors["sku"].append(
-                ValidationError(
-                    "Duplicated SKU.", ProductErrorCode.UNIQUE, params={"index": index}
+    def validate_base_fields(
+        cls, cleaned_input, duplicated_sku, errors, index_error_map, index
+    ):
+        base_fields_errors_count = 0
+        weight = cleaned_input.get("weight")
+        if weight and weight.value < 0:
```
[the remark this answers is not in the document]
Comment (SzymJ) [reply saleor/graphql/product/bulk_mutations/product_variant_bulk_create.py:476]: This is a good idea, but this change will break other mutations, so I think it should be done in separate PR.

**Review by fowczarek — APPROVED** [review]

**Review by maarcingebala — APPROVED** [review]

**Pull request merged**